src/overlays/actors/ovl_Bg_Ddan_Jd/z_bg_ddan_jd.c

```diff
@@ -31,17 +33,152 @@ const ActorInit Bg_Ddan_Jd_InitVars = {
     (ActorFunc)BgDdanJd_Update,
     (ActorFunc)BgDdanJd_Draw,
 };
-*/
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ddan_Jd/BgDdanJd_Init.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ddan_Jd/BgDdanJd_Destroy.s")
+static InitChainEntry sInitChain[] = {
+    ICHAIN_VEC3F_DIV1000(scale, 100, ICHAIN_STOP),
+};
+
+#define STATE_GO_BOTTOM 0
+#define STATE_GO_MIDDLE_FROM_BOTTOM 1
+#define STATE_GO_MIDDLE_FROM_TOP 2
+#define STATE_GO_TOP 3
+
+#define MOVE_HEIGHT_MIDDLE 140.0f
+#define MOVE_HEIGHT_TOP 700.0f
+
+#define IDLE_FRAMES 100
+
+// since stepSpeed also determines whether the platform is the shortcut to the top of the dungeon, these must be
+// different values to work correctly
+#define DEFAULT_STEP_SPEED 1
+#define SHORTCUT_STEP_SPEED 5
+
+void BgDdanJd_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgDdanJd* this = THIS;
+    s32 pad;
+    ColHeader* colHeader = NULL;
+
+    Actor_ProcessInitChain(thisx, sInitChain);
+    DynaPolyInfo_SetActorMove(thisx, DPM_PLAYER);
+    DynaPolyInfo_Alloc(&D_06003CE0, &colHeader);
+    this->dyna.dynaPolyId = DynaPolyInfo_RegisterActor(globalCtx, &globalCtx->colCtx.dyna, thisx, colHeader);
+    this->idleTimer = IDLE_FRAMES;
+    this->state = STATE_GO_BOTTOM;
+
+    // @bug: Missing check for actor.params < 0x40. This will cause inconsistent behavior if params >= 0x40 and the
+    // bound switch state is turned on while in the same room, as the shortcut behavior won't become enabled until the
+    // actor is reloaded.
+    if (Flags_GetSwitch(globalCtx, this->dyna.actor.params)) {
+        this->stepSpeed = SHORTCUT_STEP_SPEED;
+    } else {
+        this->stepSpeed = DEFAULT_STEP_SPEED;
+    }
+    this->actionFunc = BgDdanJd_Idle;
+}
+
+void BgDdanJd_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgDdanJd* this = THIS;
+    DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, this->dyna.dynaPolyId);
+}
+
+// Implements the platform's idle state
+void BgDdanJd_Idle(BgDdanJd* this, GlobalContext* globalCtx) {
+
+    if (this->idleTimer != 0) {
+        this->idleTimer--;
+    }
+
+    // if this is the platform that rises all the way to the top, and the switch state has turned on
+    if (this->stepSpeed == DEFAULT_STEP_SPEED && this->dyna.actor.params < 0x40 &&
+        Flags_GetSwitch(globalCtx, this->dyna.actor.params)) {
+        this->stepSpeed = SHORTCUT_STEP_SPEED;
+        this->state = STATE_GO_MIDDLE_FROM_BOTTOM;
+        this->idleTimer = 0;
+        this->dyna.actor.posRot.pos.y = this->dyna.actor.initPosRot.pos.y + MOVE_HEIGHT_MIDDLE;
+        func_800800F8(globalCtx, 0xBF4, -0x63, &this->dyna.actor, 0);
+    }
+    if (this->idleTimer == 0) {
+        this->idleTimer = IDLE_FRAMES;
+        if (this->state == STATE_GO_BOTTOM) {
+            this->state = STATE_GO_MIDDLE_FROM_BOTTOM;
+            this->targetHeight = this->dyna.actor.initPosRot.pos.y + MOVE_HEIGHT_MIDDLE;
```

Comment:
I would rename `targetHeight` to `targetY` since "height" in `MOVE_HEIGHT_MIDDLE` is relative to the initial Y, while targetHeight is not
(height == relative, Y == absolute)